x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/tree/Node.java

```diff
@@ -218,13 +218,28 @@ public T transformUp(Function<? super T, ? extends T> rule) {
             transformedChildren.add(next);
         }
 
-        return (childrenChanged ? replaceChildren(transformedChildren) : (T) this);
+        return (childrenChanged ? replaceEachChild(transformedChildren) : (T) this);
     }
 
     /**
-     * Replace the children of this node.
+     * Replaces each child of the current node and validates that the number of children before 
+     * and after this operation will stay the same (default invariant property of node transformations).
      */
-    public abstract T replaceChildren(List<T> newChildren);
+    public final T replaceEachChild(List<T> newChildren) {
+        // check for the invariant here once, instead of checking within every class
+        // invariant: before and after any transform the number of children should not change
+        if (newChildren.size() != children.size()) {
+            throw new QlIllegalArgumentException(
+                "Expected the same number of children [" + children.size() + "], but received [" + newChildren.size() + "]");
+        }
+        return replaceChildren(newChildren);
+    }
+
+    /**
```

Comment:
I don't see the new javadoc as an improvement over the previous doc. Why is there a need to clarify `replaceChildren` is `keeping the current properties` ?

Reply:
Removed doc.